spring-kafka/src/main/java/org/springframework/kafka/config/MethodKafkaListenerEndpoint.java

```diff
@@ -114,10 +115,26 @@ protected HandlerAdapter configureListenerAdapter(MessagingMessageListenerAdapte
 
 	/**
 	 * Create an empty {@link MessagingMessageListenerAdapter} instance.
+	 * @param messageConverter the converter (may be null).
 	 * @return the {@link MessagingMessageListenerAdapter} instance.
 	 */
-	protected MessagingMessageListenerAdapter<K, V> createMessageListenerInstance() {
-		return new MessagingMessageListenerAdapter<K, V>(this.method);
+	protected MessagingMessageListenerAdapter<K, V> createMessageListenerInstance(MessageConverter messageConverter) {
+		if (isBatchListener()) {
+			BatchMessagingMessageListenerAdapter<K, V> messageListener = new BatchMessagingMessageListenerAdapter<K, V>(
+					this.method);
```

Comment:
Because `List<ConsumerRecord<?,?>` works today (with only one entry of course).

How would we know whether `List<String>` is for a batch listener or not?

`List<String>` might the be the payload of a single message (deserialized JSON) so we need a flag (unless we add it to the annotation as you suggested). I suppose adding it to the annotation would mean the same factory could be used for batches and not batches. 

But I am inclined to keep it simple; we can always enhance it later if the community insists.